src/NServiceBus.Core/Persistence/InMemory/Gateway/InMemoryGatewayDeduplication.cs

```diff
@@ -2,63 +2,47 @@
 {
     using System;
     using System.Collections.Generic;
-    using System.Linq;
     using System.Threading.Tasks;
     using Extensibility;
     using Gateway.Deduplication;
 
     class InMemoryGatewayDeduplication : IDeduplicateMessages
     {
+        public InMemoryGatewayDeduplication(int maxSize)
+        {
+            MaxSize = maxSize;
+        }
+
         public Task<bool> DeduplicateMessage(string clientId, DateTime timeReceived, ContextBag context)
         {
-            lock (persistence)
+            lock (clientIdSet)
             {
-                var item = persistence.SingleOrDefault(m => m.Id == clientId);
-                if (item != null)
+                // Return FALSE if item EXISTS, TRUE if ADDED
+                if (clientIdSet.TryGetValue(clientId, out var existingNode)) // O(1)
                 {
+                    clientIdList.Remove(existingNode); // O(1) operation, because we got the node reference
+                    clientIdList.AddLast(existingNode); // O(1) operation
                     return TaskEx.FalseTask;
                 }
-
-                return Task.FromResult(persistence.Add(new GatewayMessage
+                else
                 {
-                    Id = clientId,
-                    TimeReceived = timeReceived
-                }));
-            }
-        }
-
-        public int DeleteDeliveredMessages(DateTime until)
```

Comment:
So was this actually ever called anywhere? Seems strange that you could just delete this without needing to modify any calling code.

Reply:
That is correct @bording :-) dead code